component/upgrade/plugins/src/main/java/org/exoplatform/platform/upgrade/plugins/AbstractGadgetToPortletPlugin.java

```diff
@@ -0,0 +1,236 @@
+                pageData.getTitle(),
+                pageData.getDescription(),
+                pageData.getWidth(),
+                pageData.getHeight(),
+                pageData.getAccessPermissions(),
+                children,
+                pageData.getOwnerType(),
+                pageData.getOwnerId(),
+                pageData.getEditPermission(),
+                pageData.isShowMaxWindow(),
+                pageData.getMoveAppsPermissions(),
+                pageData.getMoveContainersPermissions()
+        );
+    }
+
+    protected List<ComponentData> migrateComponents(List<ComponentData> components) throws Exception {
+        List<ComponentData> children = new ArrayList<>();
+        for(ComponentData ele : components) {
+            ComponentData migrated = this.migrateComponent(ele);
+            if (migrated != null) {
+                children.add(migrated);
+            }
+        }
+        return children;
+    }
+
+    protected ComponentData migrateComponent(ComponentData component) throws Exception {
+        if (component instanceof ApplicationData) {
+            ApplicationData<?> app = (ApplicationData<?>)component;
+            ApplicationState state = app.getState();
+            String contentId = this.modelDataStorage.getId(state);
+
+            if (app.getType() == ApplicationType.GADGET) {
+                LOG.info("Migrate for gadget: " + contentId);
+
+                String newContentId = null;
+                if ("Bookmark".equals(contentId)) {
+                    newContentId = "portlets/Bookmark";
+                } else if ("rssAggregator".equals(contentId)) {
+                    newContentId = "portlets/RSSReader";
+                } else if ("LoginHistory".equals(contentId)) {
+                    newContentId = "portlets/LoginHistory";
+                } else if ("FeaturedPoll".equals(contentId)) {
+                    newContentId = "portlets/FeaturedPoll";
+                }
+
+                if (newContentId != null) {
+                    LOG.info("== Replace by new portlet " + newContentId);
+                    Application<Portlet> a = Application.createPortletApplication();
+                    a.setState(new TransientApplicationState<>(newContentId, null));
+                    a.setAccessPermissions(app.getAccessPermissions().toArray(new String[0]));
+                    a.setIcon(app.getIcon());
+                    a.setShowApplicationMode(app.isShowApplicationMode());
+                    a.setShowApplicationState(app.isShowApplicationState());
+                    a.setShowInfoBar(app.isShowInfoBar());
+                    a.setTheme(app.getTheme());
+                    a.setTitle(app.getTitle());
+                    a.setDescription(app.getDescription());
+                    a.setWidth(app.getWidth());
+                    a.setHeight(app.getHeight());
+
+                    return (ApplicationData)a.build();
+                }
+                LOG.info("Migrate gadget done");
```

Comment:
If we are here, it means that the gadget is removed, right ? If so, the log should be:
```suggestion
                LOG.info("Gagdet " + contentId + " removed from page " + page);
```